enhancements/ingress/configurable-dns-loadbalancerservice.md

```diff
@@ -0,0 +1,199 @@
+available as a Day-2 operation to the cluster admin.
+
+#### As a cluster admin, when updating an ingresscontroller from managed DNS to unmanaged DNS, I want the previous DNS records to be deleted
+
+Upon updating `manageDNS` to `false` the older DNS records that were created
+must be deleted by the ingress operator.
+
+### Goals
+
+- Provide the ability to opt in for DNS management by the `cluster-ingress-operator`.
+- Recover from degraded state of the `cluster-ingress-operator` during upgrades
+  from OCP 3.x to 4.x where DNS was being managed externally,
+
+### Non-Goals
+
+- Provide a day-0 solution for cluster installations involving external DNS
+  managaments by the customer.
+
+## Proposal
+
+Currently there is partial support provided by the `installer` (in UPI installations)
+to disable the `cluster-ingress-operator` from managing the DNS cluster-wide as documented
+[here](https://github.com/openshift/installer/blob/master/docs/user/aws/install_upi.md#remove-dns-zones). 
+
+The proposed solution adds support for more fine grained control over specific
+ingresscontrollers on how they manage the wildcard DNS records associated with them.
+
+### API Extensions
+
+The IngressController CRD is extended by adding an optional `ManageDNS`
+field with type `bool`. This field will default to true.
+
+```go
+// LoadBalancerStrategy holds parameters for a load balancer.
+type LoadBalancerStrategy struct {
+    // <snip>
+
+    // manageDNS indicates if the lifecyle of the wildcard DNS record
+    // associated with the load balancer service will be managed by
+    // the ingress operator.
+    //
+	// +kubebuilder:default:=true
+	// +kubebuilder:validation:Optional 
+    // +optional
+    ManageDNS bool `json:"manageDNS"`
+}
+```
+
+Re-uses existing state `DNSManaged` of the controller to denote to denote if
+the wildcard *DNSRecord* associated with the controller is managed/un-managed. 
+
+### Implementation Details/Notes/Constraints
+
+Based on `managedDNS`, the controller decides whether to ensure creating
+the a wildcard *DNSRecord*. If the customer is updating `managedDNS` from 
+`"true"` to `"false"` the previously created DNSRecord will be deleted.
+
+The support provided by the installer and this new field aren't entriely
+mutually exclusive i.e. if customer has opted to disable DNS management cluster
+wide (as mentioned under UPI based installations), this additional flag is of
+no value and need not be set since *DNSRecord* created is silently reconciled 
+by the dns controller.
+
+### Risks and Mitigations
+
+The manual creation of the required DNS records will need to be done prior to
+making this change as the canary controller will try to ensure the if the ingress
+domains are reachable, otherwise the `ingress` *ClusterOperator* results in a 
+degraded state. This needs to be documented so as to avoid unecessary support cases.
+
+### Drawbacks
+
+There is an inconsistency in how the existing support is added to disable
```

Comment:
I find it quite consistent. Before the proposed solution the ingress controller was always managing `DNSRecord` CRs for its domain. So it keeps managing them fulfilling its contract with the customer no matter what. The fact that the zone may not exist has to be handled but it doesn't free the ingress controller from its contract. Only the new API field you are adding may do this.

Reply:
So the reason why I feel it is inconsistent is because, once this new field is introduced, some controllers will not have any *DNSRecord*s associated with them at all compared to the scenario where the customer setup the installer without zones but still the ingresscontrollers have *DNSRecords* associated with them just to denote the association between the domain and the target external IP. (although since it's using a fake provider, the association isn't actually ensured by the operator).

Reply:
I think I see what you mean. In my opinion this comes from the unprecise wording in the UPI markdown docs:

> If you don't want [the ingress operator](https://github.com/openshift/cluster-ingress-operator) to manage DNS records on your behalf, remove the privateZone and publicZone sections from the DNS configuration:

They are putting it like the cluster ingress operator explicitly stops doing any DNS related things if no cluster zone is present. While in reality the operator keeps doing what it was doing, but since there is no zone - no actual DNS records is created. The final result is the same, I agree, but the intention is different. I see it as a sort of separation of concerns: zones are managed in one place, ingress operator does its job in another place; and only this lonely doc puts an explicit link between these two.

Well, overall, I  don't think it's a real drawback, it's maybe a slightly unprecise doc but I don't think it's something we should worry about for this EP.